File: benchmarks/compute/three-mm/C/3mm.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <time.h>

#define SMALL 0
#define STANDARD 1
#define LARGE 2

// ---------- 可选数据规模 ----------
#if !defined(DATASET)
#define DATASET SMALL
#endif

#if DATASET == SMALL
#define NI 250
#define NJ 300
#define NK 350
#define NL 400
#define NM 450
#elif DATASET == LARGE
#define NI 2000
#define NJ 2200
#define NK 2400
#define NL 2600
#define NM 2800
#else
#define NI 500
#define NJ 600
#define NK 700
#define NL 800
#define NM 900
#endif

// ---------- 初始化随机矩阵 ----------
void init_array(int ni, int nj, int nk, int nl, int nm,
               double A[ni][nk], double B[nk][nj],
               double C[nj][nm], double D[nm][nl],
               unsigned int seed) {
    srand(seed);
    for (int i = 0; i < ni; i++)
        for (int k = 0; k < nk; k++)
            A[i][k] = ((double)rand() / RAND_MAX) * 10;
    
    for (int k = 0; k < nk; k++)
        for (int j = 0; j < nj; j++)
            B[k][j] = ((double)rand() / RAND_MAX) * 10;
    
    for (int j = 0; j < nj; j++)
        for (int m = 0; m < nm; m++)
            C[j][m] = ((double)rand() / RAND_MAX) * 10;
    
    for (int m = 0; m < nm; m++)
        for (int l = 0; l < nl; l++)
            D[m][l] = ((double)rand() / RAND_MAX) * 10;
}

// ---------- 打印部分输出矩阵（防止被编译器优化） ----------
void print_matrix(int ni, int nl, double G[ni][nl]) {
    for (int i = 0; i < ni && i < 8; i++) {
        for (int l = 0; l < nl && l < 8; l++) {
            printf("%0.2f ", G[i][l]);
        }
        printf("\n");
    }
}

// ---------- 主计算过程 ----------
void compute_3mm(int ni, int nj, int nk, int nl, int nm,
                double A[ni][nk], double B[nk][nj],
                double C[nj][nm], double D[nm][nl],
                double E[ni][nj], double F[nj][nl],
                double G[ni][nl]) {
    // E = A * B
    for (int i = 0; i < ni; i++) {
        for (int j = 0; j < nj; j++) {
            E[i][j] = 0.0;
            for (int k = 0; k < nk; k++) {
                E[i][j] += A[i][k] * B[k][j];
            }
        }
    }
    
    // F = C * D
    for (int j = 0; j < nj; j++) {
        for (int l = 0; l < nl; l++) {
            F[j][l] = 0.0;
            for (int m = 0; m < nm; m++) {
                F[j][l] += C[j][m] * D[m][l];
            }
        }
    }
    
    // G = E * F
    for (int i = 0; i < ni; i++) {
        for (int l = 0; l < nl; l++) {
            G[i][l] = 0.0;
            for (int j = 0; j < nj; j++) {
                G[i][l] += E[i][j] * F[j][l];
            }
        }
    }
}

// ---------- 主函数 ----------
int main(int argc, char** argv) {
    static double A[NI][NK];
    static double B[NK][NJ];
    static double C[NJ][NM];
    static double D[NM][NL];
    static double E[NI][NJ];
    static double F[NJ][NL];
    static double G[NI][NL];

    unsigned int seed = (argc > 1) ? atoi(argv[1]) : 42;

    init_array(NI, NJ, NK, NL, NM, A, B, C, D, seed);

    clock_t start = clock();
    compute_3mm(NI, NJ, NK, NL, NM, A, B, C, D, E, F, G);
    clock_t end = clock();

    printf("Finished 3mm calculation in %.3f seconds\n",
           (double)(end - start) / CLOCKS_PER_SEC);

    print_matrix(NI, NL, G); // 打印前8x8子矩阵
    return 0;
}

File: benchmarks/compute/three-mm/Python/3mm.py
import sys
import time
import random

# ---------- 可选数据规模 ----------
SMALL = 0
STANDARD = 1
LARGE = 2

DATASET = SMALL

if DATASET == SMALL:
    NI, NJ, NK, NL, NM = 250, 300, 350, 400, 450
elif DATASET == LARGE:
    NI, NJ, NK, NL, NM = 2000, 2200, 2400, 2600, 2800
else:
    NI, NJ, NK, NL, NM = 500, 600, 700, 800, 900

# ---------- 初始化随机矩阵 ----------
def init_array(seed):
    random.seed(seed)
    A = [[random.random() * 10 for _ in range(NK)] for _ in range(NI)]
    B = [[random.random() * 10 for _ in range(NJ)] for _ in range(NK)]
    C = [[random.random() * 10 for _ in range(NM)] for _ in range(NJ)]
    D = [[random.random() * 10 for _ in range(NL)] for _ in range(NM)]
    return A, B, C, D

# ---------- 打印部分输出矩阵 ----------
def print_matrix(G):
    for i in range(min(8, len(G))):
        row = G[i]
        for l in range(min(8, len(row))):
            print(f"{row[l]:.2f} ", end="")
        print()

# ---------- 主计算过程 ----------
def compute_3mm(A, B, C, D):
    # E = A * B
    E = [[0.0 for _ in range(NJ)] for _ in range(NI)]
    for i in range(NI):
        for j in range(NJ):
            for k in range(NK):
                E[i][j] += A[i][k] * B[k][j]
    
    # F = C * D
    F = [[0.0 for _ in range(NL)] for _ in range(NJ)]
    for j in range(NJ):
        for l in range(NL):
            for m in range(NM):
                F[j][l] += C[j][m] * D[m][l]
    
    # G = E * F
    G = [[0.0 for _ in range(NL)] for _ in range(NI)]
    for i in range(NI):
        for l in range(NL):
            for j in range(NJ):
                G[i][l] += E[i][j] * F[j][l]
    
    return G

# ---------- 主函数 ----------
if __name__ == "__main__":
    seed = int(sys.argv[1]) if len(sys.argv) > 1 else 42
    A, B, C, D = init_array(seed)
    
    start = time.time()
    G = compute_3mm(A, B, C, D)
    end = time.time()
    
    print(f"Finished 3mm calculation in {end - start:.3f} seconds")
    print_matrix(G)
